File: python/src/smooai_audit/hash.py
"""SHA-256 hash chain over audit events.

Each event's ``hash_current`` is ``SHA-256(canonical-JSON(event minus
hash_current))``; the chain links via ``hash_previous`` = the prior event's
``hash_current``. :func:`verify_chain` replays that: recompute every hash and
confirm each event's ``hash_previous`` equals the previous event's
``hash_current``.
"""

from __future__ import annotations

import hashlib
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from enum import Enum
from typing import Any

from .canonical import canonical_json
from .schema import AuditEvent


def compute_event_hash(event: AuditEvent | Mapping[str, Any]) -> str:
    """Return the lowercase hex SHA-256 of ``event`` minus its own
    ``hashCurrent`` field. Accepts an :class:`AuditEvent` or a raw wire mapping
    (camelCase keys, as decoded from JSON)."""
    if isinstance(event, AuditEvent):
        payload: Any = event.model_dump(by_alias=True, exclude_unset=True, mode="json")
    else:
        payload = dict(event)
    payload.pop("hashCurrent", None)
    return hashlib.sha256(canonical_json(payload).encode("utf-8")).hexdigest()


def build_hash_chain(events: list[AuditEvent], previous_hash: str | None = None) -> list[AuditEvent]:
    """Seal an ordered list of events into a chain: stamp each with its
    ``hash_previous`` (the prior event's ``hash_current``, or ``previous_hash``
    for the first) and its freshly computed ``hash_current``. Returns new sealed
    copies; inputs are not mutated."""
    sealed: list[AuditEvent] = []
    prev = previous_hash
    for event in events:
        # Omit hash_previous entirely on the first event of a chain (JS undefined,
        # not null) — only stamp it once there is a prior hash to link to.
        update: dict[str, Any] = {"hash_current": None}
        if prev is not None:
            update["hash_previous"] = prev
        with_prev = event.model_copy(update=update)
        current = compute_event_hash(with_prev)
        sealed.append(with_prev.model_copy(update={"hash_current": current}))
        prev = current
    return sealed


class VerifyFailureCode(str, Enum):
    """Why a chain failed to verify.

    These codes are the cross-language contract — every SDK returns the same code
    for the same corruption, asserted by ``chainFixtures`` in
    ``spec/parity-corpus.json``.
    """

    HASH_PREVIOUS_MISMATCH = "hash_previous_mismatch"
    """An event's ``hash_previous`` is not the prior event's ``hash_current`` —
    the LINK is wrong: a reorder, a deletion, a truncated head, a rewritten link."""

    HASH_CURRENT_MISMATCH = "hash_current_mismatch"
    """The event's own content no longer hashes to its stored ``hash_current`` —
    the event BODY was edited after sealing."""


@dataclass(frozen=True)
class ChainVerification:
    """The verdict from :func:`verify_chain`. Falsy when the chain is broken."""

    ok: bool
    broken_at: int | None = None
    """Index of the first event that failed. ``None`` when ``ok``."""
    code: VerifyFailureCode | None = None
    """Why it failed. ``None`` when ``ok``."""

    def __bool__(self) -> bool:
        return self.ok


def _chain_field(event: AuditEvent | Mapping[str, Any], attr: str, wire_key: str) -> str | None:
    """Read a chain field from either a typed event or a raw wire mapping."""
    if isinstance(event, AuditEvent):
        return getattr(event, attr)
    value = event.get(wire_key)
    return value if isinstance(value, str) else None


def verify_chain(
    events: Sequence[AuditEvent | Mapping[str, Any]],
    genesis_previous_hash: str | None = None,
) -> ChainVerification:
    """Verify an ordered chain: recompute every ``hash_current`` and confirm each
    ``hash_previous`` matches the prior event's ``hash_current``.

    ``genesis_previous_hash`` is the hash the FIRST event must link to — pass the
    chain head you already have when verifying a slice that continues an existing
    chain. Leave it ``None`` only when ``events`` starts at the true beginning of
    the chain (first event of the org's day), where ``hash_previous`` must be
    absent.

    **What replay cannot see:** removing events from the TAIL leaves a chain that
    still verifies — every remaining link is genuine. Detecting that needs an
    external anchor (a stored chain head, an expected count) compared against the
    last event's ``hash_current``. An ``ok`` verdict means "nothing here was
    altered", not "nothing is missing"; the corpus pins this as an explicit
    fixture so the limit stays visible.
    """
    previous = genesis_previous_hash
    for index, event in enumerate(events):
        if _chain_field(event, "hash_previous", "hashPrevious") != previous:
            return ChainVerification(ok=False, broken_at=index, code=VerifyFailureCode.HASH_PREVIOUS_MISMATCH)
        recomputed = compute_event_hash(event)
        if _chain_field(event, "hash_current", "hashCurrent") != recomputed:
            return ChainVerification(ok=False, broken_at=index, code=VerifyFailureCode.HASH_CURRENT_MISMATCH)
        previous = recomputed
    return ChainVerification(ok=True)

File: rust/audit/src/hash.rs
//! Per-event SHA-256 + the per-org-per-day hash chain.
//!
//! `hashCurrent = SHA-256(canonical-JSON(event minus hashCurrent))`, linked via
//! `hashPrevious`. Byte-verified against `spec/parity-corpus.json`.

use sha2::{Digest, Sha256};

use crate::canonical::canonical_json;
use crate::schema::AuditEvent;

/// Lowercase-hex SHA-256 of the canonical JSON of `value`.
///
/// `value` MUST be the event object **without** `hashCurrent` (and with
/// `hashPrevious` already set to the chain head, or omitted on the first event
/// of a day). See [`AuditEvent::event_for_hash`].
pub fn compute_event_hash(value: &serde_json::Value) -> String {
    let digest = Sha256::digest(canonical_json(value).as_bytes());
    // Lowercase hex without pulling in the `hex` crate.
    let mut out = String::with_capacity(digest.len() * 2);
    for byte in digest {
        out.push_str(&format!("{byte:02x}"));
    }
    out
}

/// Fold events into a per-org-per-day hash chain, stamping each with its
/// `hashPrevious` (the prior event's hash) and `hashCurrent`.
///
/// `chain_head` is the hash of the last event already in today's chain, or
/// `None` for the first event of the day (which gets no `hashPrevious`). Returns
/// the events with both hash fields set, in order.
pub fn build_hash_chain(events: Vec<AuditEvent>, chain_head: Option<String>) -> Vec<AuditEvent> {
    let mut head = chain_head;
    events
        .into_iter()
        .map(|mut event| {
            event.hash_previous = head.clone();
            let hash = event.compute_hash();
            event.hash_current = Some(hash.clone());
            head = Some(hash);
            event
        })
        .collect()
}

/// Why a chain failed to verify.
///
/// These codes are the cross-language contract — every SDK returns the same code
/// for the same corruption, asserted by `chainFixtures` in
/// `spec/parity-corpus.json`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum VerifyFailureCode {
    /// An event's `hashPrevious` is not the prior event's `hashCurrent` — the
    /// LINK is wrong: a reorder, a deletion, a truncated head, a rewritten link.
    HashPreviousMismatch,
    /// The event's own content no longer hashes to its stored `hashCurrent` —
    /// the event BODY was edited after sealing.
    HashCurrentMismatch,
}

/// The verdict from [`verify_chain`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ChainVerification {
    /// Every link recomputed and matched.
    Ok,
    /// The first event that failed, and why.
    Broken { broken_at: usize, code: VerifyFailureCode },
}

impl ChainVerification {
    /// `true` when the chain verified intact.
    pub fn is_ok(&self) -> bool {
        matches!(self, ChainVerification::Ok)
    }
}

/// Verify an ordered chain: recompute every `hashCurrent` and confirm each
/// `hashPrevious` matches the prior event's `hashCurrent`.
///
/// `genesis_previous_hash` is the hash the FIRST event must link to — pass the
/// chain head you already have when verifying a slice that continues an existing
/// chain. `None` means `events` starts at the true beginning of the chain (first
/// event of the org's day), where `hashPrevious` must be absent.
///
/// **What replay cannot see:** removing events from the TAIL leaves a chain that
/// still verifies — every remaining link is genuine. Detecting that needs an
/// external anchor (a stored chain head, an expected count) compared against the
/// last event's `hashCurrent`. [`ChainVerification::Ok`] means "nothing here was
/// altered", not "nothing is missing"; the corpus pins this as an explicit
/// fixture so the limit stays visible.
pub fn verify_chain(events: &[AuditEvent], genesis_previous_hash: Option<&str>) -> ChainVerification {
    let mut previous: Option<String> = genesis_previous_hash.map(str::to_owned);
    for (index, event) in events.iter().enumerate() {
        if event.hash_previous != previous {
            return ChainVerification::Broken {
                broken_at: index,
                code: VerifyFailureCode::HashPreviousMismatch,
            };
        }
        let recomputed = event.compute_hash();
        if event.hash_current.as_deref() != Some(recomputed.as_str()) {
            return ChainVerification::Broken {
                broken_at: index,
                code: VerifyFailureCode::HashCurrentMismatch,
            };
        }
        previous = Some(recomputed);
    }
    ChainVerification::Ok
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn hash_is_64_hex_lowercase() {
        let h = compute_event_hash(&json!({ "a": 1 }));
        assert_eq!(h.len(), 64);
        assert!(h.chars().all(|c| c.is_ascii_hexdigit() && !c.is_ascii_uppercase()));
    }

    #[test]
    fn chain_links_previous_to_prior_current() {
        let base = AuditEvent {
            id: "id".into(),
            organization_id: "org".into(),
            actor_type: crate::schema::ActorType::System,
            actor_id: "sys".into(),
            actor_email: None,
            action: "x.y".into(),
            resource: crate::schema::AuditResource {
                type_: "t".into(),
                id: "i".into(),
            },
            outcome: crate::schema::Outcome::Success,
            reason: None,
            session_id: None,
            conversation_id: None,
            ip_address: None,
            user_agent: None,
            geo_country: None,
            diff: None,
            metadata: Default::default(),
            timestamp: "2026-01-01T00:00:00.000Z".into(),
            hash_previous: None,
            hash_current: None,
        };
        let mut second = base.clone();
        second.id = "id2".into();
        let chain = build_hash_chain(vec![base, second], None);
        assert_eq!(chain[0].hash_previous, None, "first event has no hashPrevious");
        assert_eq!(chain[1].hash_previous, chain[0].hash_current, "link to prior current");
        assert!(chain[0].hash_current.is_some());
    }
}
